Tests fail then `busted` is already installed in a "user" tree for a Lua version different from 5.1

### Problem

The problem is outlined in the title, details are under "Steps to reproduce". I also found an issue, which may be somewhat related to this one https://github.com/neovim/neovim/issues/23814

### Steps to reproduce

Install Lua 5.3 and luarocks via a system package manager (in my case, `sudo pacman -S lua53 luarocks`).

Make sure `~/.luarocks` does not exist at that point, to start cleanly.

Set Lua 5.3 to be the default version via `luarocks config lua_version 5.3` and enable `local_by_default` via `luarocks config local_by_default true`. This should result in the following `~/.luarocks/default-lua-version.lua`:

```lua
return "5.3"
```

And `~/.luarocks/default-lua-version.lua`:

```lua
local_by_default = true
lua_version = "5.3"
```

Install busted via `luarocks install busted`

Run `make distclean` and then `make test` in a Neovim repo root.

I'm seeing the following output:

```
-- Using NVIM_VERSION: v0.10.0-dev-952+gb1cfb299d
[206/396] Performing download step (download, verify and extract) for 'luarocks'
-- Downloading...
   dst='/home/faergeek/projects/neovim/build/build/downloads/luarocks/v3.9.2.tar.gz'
   timeout='none'
   inactivity timeout='none'
-- Using src='https://github.com/luarocks/luarocks/archive/v3.9.2.tar.gz'
-- verifying file...
       file='/home/faergeek/projects/neovim/build/build/downloads/luarocks/v3.9.2.tar.gz'
-- Downloading... done
-- extracting...
     src='/home/faergeek/projects/neovim/build/build/downloads/luarocks/v3.9.2.tar.gz'
     dst='/home/faergeek/projects/neovim/build/build/src/luarocks'
-- extracting... [tar xfz]
-- extracting... [analysis]
-- extracting... [rename]
-- extracting... [clean up]
-- extracting... done
[227/396] Performing configure step for 'luarocks'

Configuring LuaRocks version 3.9.2...

Lua version detected: 5.1
Lua interpreter found: /home/faergeek/projects/neovim/.deps/usr/bin/luajit
lua.h found: /home/faergeek/projects/neovim/.deps/usr/include/luajit-2.1/lua.h
unzip found in PATH: /usr/bin

Done configuring.

LuaRocks will be installed at......: /home/faergeek/projects/neovim/build/usr
LuaRocks will install rocks at.....: /home/faergeek/projects/neovim/build/usr
LuaRocks configuration directory...: /home/faergeek/projects/neovim/build/usr/etc/luarocks
Using Lua from.....................: /home/faergeek/projects/neovim/.deps/usr
Lua include directory..............: /home/faergeek/projects/neovim/.deps/usr/include/luajit-2.1

* Type make and make install:
  to install to /home/faergeek/projects/neovim/build/usr as usual.
* Type make bootstrap:
  to install LuaRocks into /home/faergeek/projects/neovim/build/usr as a rock.

[378/396] Performing install step for 'luarocks'
make[1]: Entering directory '/home/faergeek/projects/neovim/build/build/src/luarocks'
mkdir -p "build"
mkdir -p .luarocks
cp ./build/config-5.1.lua .luarocks/config-5.1.lua
rm -f src/luarocks/core/hardcoded.lua
echo "#!/bin/sh" > luarocks
echo "unset LUA_PATH LUA_PATH_5_2 LUA_PATH_5_3 LUA_PATH_5_4 LUA_CPATH LUA_CPATH_5_2 LUA_CPATH_5_3 LUA_CPATH_5_4" >> luarocks
echo 'LUAROCKS_SYSCONFDIR="/home/faergeek/projects/neovim/build/usr/etc/luarocks" LUA_PATH="/home/faergeek/projects/neovim/build/build/src/luarocks/src/?.lua;;" exec "/home/faergeek/projects/neovim/.deps/usr/bin/luajit" "/home/faergeek/projects/neovim/build/build/src/luarocks/src/bin/luarocks" --project-tree="/home/faergeek/projects/neovim/build/build/src/luarocks/lua_modules" "$@"' >> luarocks
chmod +rx ./luarocks
./luarocks init

Initializing project 'luarocks' for Lua 5.3 ...
-----------------------------------------------

Checking your Lua installation ...
Adding entries to .gitignore ...
Preparing ./.luarocks/ ...
Wrote .luarocks/config-5.3.lua
Preparing ./lua_modules/ ...
./luarocks already exists. Not overwriting it!
Preparing ./lua for version 5.3...
mkdir -p "/home/faergeek/projects/neovim/build/usr/etc/luarocks"
install -m 644 "build/config-5.1.lua" "/home/faergeek/projects/neovim/build/usr/etc/luarocks/config-5.1.lua"
./luarocks make --tree="/home/faergeek/projects/neovim/build/usr"

No existing manifest. Attempting to rebuild...
luarocks 3.9.2-1 is now installed in /home/faergeek/projects/neovim/build/usr (license: MIT)

make[1]: Leaving directory '/home/faergeek/projects/neovim/build/build/src/luarocks'
[394/396] Generating usr/bin/luacheck

luacheck 1.1.0-1 depends on lua >= 5.1 (5.3-1 provided by VM)
luacheck 1.1.0-1 depends on argparse >= 0.6.0 (0.7.1-1 installed)
luacheck 1.1.0-1 depends on luafilesystem >= 1.6.3 (1.8.0-1 installed)
luacheck 1.1.0-1 is now installed in /home/faergeek/.luarocks (license: MIT)

[395/396] Generating usr/bin/busted

busted 2.1.1-1 depends on lua >= 5.1 (5.3-1 provided by VM)
busted 2.1.1-1 depends on lua_cliargs 3.0 (3.0-2 installed)
busted 2.1.1-1 depends on luafilesystem >= 1.5.0 (1.8.0-1 installed)
busted 2.1.1-1 depends on luasystem >= 0.2.0 (0.2.1-0 installed)
busted 2.1.1-1 depends on dkjson >= 2.1.0 (2.6-1 installed)
busted 2.1.1-1 depends on say >= 1.4-1 (1.4.1-3 installed)
busted 2.1.1-1 depends on luassert >= 1.9.0-1 (1.9.0-1 installed)
busted 2.1.1-1 depends on lua-term >= 0.1 (0.7-1 installed)
busted 2.1.1-1 depends on penlight >= 1.3.2 (1.13.1-1 installed)
busted 2.1.1-1 depends on mediator_lua >= 1.1.1 (1.1.2-0 installed)
busted 2.1.1-1 is now installed in /home/faergeek/.luarocks (license: MIT <http://opensource.org/licenses/MIT>)

[395/396] cd /home/faergeek/projects/neovim/build/test && /usr/bin/cmake -D N...US_CI= -D CI_BUILD=OFF -P /home/faergeek/projects/neovim/cmake/RunTests.cmake
-- Tests exited non-zero: 1
-- Output to stderr:
/home/faergeek/projects/neovim/test/busted_runner.lua:7: module 'busted.runner' not found:
	no field package.preload['busted.runner']
	no file '/home/faergeek/projects/neovim/build/usr/share/lua/5.1/busted/runner.lua'
	no file '/home/faergeek/projects/neovim/build/usr/share/lua/5.1/busted/runner/init.lua'
	no file './busted/runner.lua'
	no file '/home/faergeek/projects/neovim/.deps/usr/share/luajit-2.1/busted/runner.lua'
	no file '/usr/local/share/lua/5.1/busted/runner.lua'
	no file '/usr/local/share/lua/5.1/busted/runner/init.lua'
	no file '/home/faergeek/projects/neovim/.deps/usr/share/lua/5.1/busted/runner.lua'
	no file '/home/faergeek/projects/neovim/.deps/usr/share/lua/5.1/busted/runner/init.lua'
	no file '/home/faergeek/projects/neovim/build/usr/lib/lua/5.1/busted/runner.so'
	no file './busted/runner.so'
	no file '/usr/local/lib/lua/5.1/busted/runner.so'
	no file '/home/faergeek/projects/neovim/.deps/usr/lib/lua/5.1/busted/runner.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
	no file '/home/faergeek/projects/neovim/build/usr/lib/lua/5.1/busted.so'
	no file './busted.so'
	no file '/usr/local/lib/lua/5.1/busted.so'
	no file '/home/faergeek/projects/neovim/.deps/usr/lib/lua/5.1/busted.so'
	no file '/usr/local/lib/lua/5.1/loadall.so'
stack traceback:
	[C]: in function 'require'
	/home/faergeek/projects/neovim/test/busted_runner.lua:7: in main chunk

CMake Error at /home/faergeek/projects/neovim/cmake/RunTests.cmake:103 (message):
  functional tests failed with error: 1


FAILED: test/CMakeFiles/functionaltest /home/faergeek/projects/neovim/build/test/CMakeFiles/functionaltest
cd /home/faergeek/projects/neovim/build/test && /usr/bin/cmake -D NVIM_PRG=/home/faergeek/projects/neovim/build/bin/nvim -D WORKING_DIR=/home/faergeek/projects/neovim -D BUSTED_OUTPUT_TYPE=nvim -D TEST_DIR=/home/faergeek/projects/neovim/test -D BUILD_DIR=/home/faergeek/projects/neovim/build -D DEPS_INSTALL_DIR=/home/faergeek/projects/neovim/build/usr -D TEST_TYPE=functional -D CIRRUS_CI= -D CI_BUILD=OFF -P /home/faergeek/projects/neovim/cmake/RunTests.cmake
ninja: build stopped: subcommand failed.
make: *** [Makefile:135: functionaltest] Error 1
```

Notice the line saying that starting from `./luarocks init` Lua 5.3 is being used instead of 5.1:

```
Initializing project 'luarocks' for Lua 5.3 ..."
```

Looks like it builds busted for Lua 5.3 only because there already exists busted for that version. Then when running tests it looks for busted luarocks for 5.1 and fails because didn't ever exist.

I was able to work around it by applying the following patch:

```patch
diff --git a/cmake/BuildLuarocks.cmake b/cmake/BuildLuarocks.cmake
index c5e08d2d7..ca38bfc3b 100644
--- a/cmake/BuildLuarocks.cmake
+++ b/cmake/BuildLuarocks.cmake
@@ -82,7 +82,7 @@ function(Download ROCK VER)
     set(OUTPUT ${ROCKS_DIR}/${ROCK})
   endif()
   add_custom_command(OUTPUT ${OUTPUT}
-    COMMAND ${CMAKE_COMMAND} -E env "${PATH}" ${LUAROCKS_BINARY} build ${ROCK} ${VER} ${LUAROCKS_BUILDARGS}
+    COMMAND ${CMAKE_COMMAND} -E env "${PATH}" ${LUAROCKS_BINARY} --lua-version 5.1 build ${ROCK} ${VER} ${LUAROCKS_BUILDARGS}
     DEPENDS ${CURRENT_DEP})
   add_custom_target(${ROCK} DEPENDS ${OUTPUT})
   set(CURRENT_DEP ${ROCK} PARENT_SCOPE)
```

But I'm pretty sure that's not the proper solution.

### Expected behavior

I suppose that better behavior would be to first check if all required rocks of appropriate versions already exist in the system for Lua 5.1. And if not, then build appropriate versions in a separate tree under `.deps`. Not sure if there's an easy way to do this though. But that would fix the following issue as well https://github.com/neovim/neovim/issues/23814

Another solution would be to just always build all the required rocks in a local tree under `.deps` and never look at or touch rocks in other trees.

### Neovim version (nvim -v)

v0.10.0-dev-952+gb1cfb299d

### Vim (not Nvim) behaves the same?

n/a

### Operating system/version

Arch Linux

### Terminal name/version

alacritty 0.12.2 (9d9982df)

### $TERM environment variable

alacritty

### Installation

build from commit b1cfb299df2ef412339f594173ed23c75c090c8a

--- a/cmake/BuildLuarocks.cmake
+++ b/cmake/BuildLuarocks.cmake
@@ -31,7 +31,7 @@ if(UNIX)
 
   set(LUAROCKS_CONFIGURE_COMMAND ${DEPS_BUILD_DIR}/src/luarocks/configure
       --prefix=${DEPS_INSTALL_DIR} --force-config ${LUAROCKS_OPTS})
-  set(LUAROCKS_INSTALL_COMMAND ${MAKE_PRG} -j1 bootstrap)
+  set(LUAROCKS_INSTALL_COMMAND ${MAKE_PRG} -j1 install)
 elseif(MSVC OR MINGW)
   if(MINGW)
     set(COMPILER_FLAG /MW)

